### PR TITLE
Added swagger for private link resource API

### PR DESCRIPTION
**Pull request by deymadhumanti**
…ults

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Review by anthony-c-martin — CHANGES_REQUESTED** [review]

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (amchandn) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: @deymadhumanti , could you please add description and readonly property for this as well?

Comment (amchandn) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: @deymadhumanti , please have description and readonly property for this as well?

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (amchandn) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: @deymadhumanti , please have description and readonly property for this as well?

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (amchandn) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: @deymadhumanti , please have description and readonly property for this as well?

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (amchandn) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: @deymadhumanti , please have description and readonly property for this as well?

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (amchandn) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: @deymadhumanti , please have description and readonly property for this as well?

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (amchandn) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: @deymadhumanti , please have description and readonly property for this as well?

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (amchandn) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Hi Anthony, Madhumanti has made the required changes in this PR for the same

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (deymadhumanti) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Done

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (deymadhumanti) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Done

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (deymadhumanti) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Done

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (pilor) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: >CheckNameAvailabilityResult [](start = 37, length = 27)

I believe this is a breaking change but I'll let the SDK reviewers comment

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (pilor) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: >It is to be returned to RRP from BMS as response of PUT PECP and PUT PEC calls [](start = 117, length = 78)

these acronyms don't mean anything to me and won't to the customer either

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (pilor) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: everything but "id", "name", and "type" should be inside a "properties" object.

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (amchandn) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Yes @pilor ,  the earlier schema for response was erronous and we are fixing it with this pull request. It is in accordance with format mentioned at - https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#check-name-availability-requests

We didn't publish an SDK with the erroneous swagger, hence it shouldn't cause an issue after checking in this PR

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (pilor) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: >"PrivateEndpointConnection" [](start = 4, length = 27)

you should have a child resource called privateEndpointConnections that exposes the full set of private endpoint properties.  See other RPs (i.e. cosmos db, key vault) for examples

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (pilor) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: >"Enabled [](start = 12, length = 8)

enabled but no disabled?  Better to be explicit about it I think.  The description should also be very explicit about what this means.  In all other scenarios I've seen enabling private endpoints enables it for the entire resource's operations

Comment (pilor) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: actually, if this is readonly what is the purpose of these two properties?

**Review by pilor — CHANGES_REQUESTED** [review]
see comments

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (sumitmal2711) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: This is actually not an arm resource proxy/tracked. Network team expects these to be in this format for RPs onboarding to Private Endpoint. For eg: See already published documentation for other onboarded RP:
https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.management.eventgrid.models.privatelinkresource?view=azure-dotnet-preview

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (deymadhumanti) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Removed this line from description

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (deymadhumanti) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: We already  have a property called PrivateEndpointConnections which exposes a full list of PrivateEndpointConnectionVaultProperties. This PrivateEndpointConnectionVaultProperties internally contains two properties : PrivateEndpointConnection and id.For each PrivateEndpointConnection’s properties, we contain only the id of the privateEndpoint as specified in section 6.5.2 of  NRP Private Endpoint Design Doc (https://microsoft.sharepoint.com/:w:/r/teams/WAG/AzureNetworking/_layouts/15/Doc.aspx?sourcedoc=%7B64F10150-563A-4715-9476-FFA24F9BD42E%7D&file=%5BNRP%5D%20Private%20Endpoint%20Design%203.docx&wdLOR=cC8C98543-7B1C-470B-A557-E1E3501F6F9B&action=default&mobileredirect=true)

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (deymadhumanti) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: We have kept ‘None’ instead of disabled as it’s the default state that the customer experiences when PrivateEndpoint scenarios haven’t been explicitly enabled.

The ARM tracked resource in this case – Vaults, is used in two scenarios – backup and siteRecovery. Both scenarios have different Groupids and different implications of PrivateEndpoint being enabled, hence having PrivateEndpointForBackup enabled will not imply PrivateEndpointForSiteRecovery to be enabled as well.

The customer isn’t supposed to set PrivateEndpointStateForBackup or PrivateEndpointStateForSiteRecovery values himself/herself. These get automatically set by the RP when there exists a privateEndpointConnection in the vault for the respective groupId in an ‘Approved’ state

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (pilor) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: You are linking to the .NET class, not the API.  They are using x-ms-client-flatten to flatten the resulting generated SDK.  The API does have it in a property bag.  See line 7309 in specification\eventgrid\resource-manager\Microsoft.EventGrid\preview\2020-04-01-preview\EventGrid.json.  This is the billionth review of these APIs, the contract is well known

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (pilor) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Please follow the example set by other RPs.  This part is correct but PrivateEndpointConnections should also be exposed as a child resource.  You should never have something with a resource ID and provisioningState that doesn't have a corresponding GET at that resource ID (and a GET collection for all of that type at the scope).  See cosmos DB, key vault, etc... for an example

Comment (deymadhumanti) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Hi, we have added GET for this PrivateEndpointConnection resource. Please check the below PR: 
https://github.com/Azure/azure-rest-api-specs/pull/8587

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (amchandn) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Thanks Chris for the explanation. We have fixed the PrivateLinkResources API response as you suggested.

**Review by pilor — APPROVED** [review]
:shipit:

**Comment by ArcturusZhang** [comment]
Hi @deymadhumanti please resolve the CI failures (Spell check and prettier). And as for the breaking changes check, could you provide a justification of this? Or it is only a false alarm?

Hi @ChenTanyi would you please take a look at the Java generation  failure?

**Comment by ChenTanyi** [comment]
The generation error is due to `privateLinkResources`, the java will make auto collections for every resource with conflict with https://github.com/Azure/azure-rest-api-specs/pull/8655/files#diff-02ce4ce9148599d2d4ec2600b9c4626cR1165. @weidongxu-microsoft Is it possible for us to generate without collections if there is already a collections.

**Comment by weidongxu-microsoft** [comment]
@ChenTanyi No idea, you know as much as I know about v2 generator.

My suggestion is to try another name for "PrivateLinkResources" definition, like "PrivateLinkResourceListResult". E.g. https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2019-08-01-preview/privateLinkResources.json#L117

**Comment by ChenTanyi** [comment]
> @ChenTanyi No idea, you know as much as I know about v2 generator.
> 
> My suggestion is to try another name for "PrivateLinkResources" definition, like "PrivateLinkResourceListResult". E.g. https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2019-08-01-preview/privateLinkResources.json#L117

@deymadhumanti Could you change for this name, if you want to update Java sdk. Thanks!

**Comment by deymadhumanti** [comment]
> > @ChenTanyi No idea, you know as much as I know about v2 generator.
> > My suggestion is to try another name for "PrivateLinkResources" definition, like "PrivateLinkResourceListResult". E.g. https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2019-08-01-preview/privateLinkResources.json#L117
> 
> @deymadhumanti Could you change for this name, if you want to update Java sdk. Thanks!

Hi  @ChenTanyi , Can we use "PrivateLinkResourceList" for PrivateLinkResources. Please confirm. I will make the json change accordingly,

**Comment by ChenTanyi** [comment]
@deymadhumanti Yes, it could be `PrivateLinkResourceList`. 

It could be anything you like which could be separated from other resources and their plurality.
For example, the java sdk will automatically collect all `privateLinkResource` into `privateLinkResources`. And so as others.

**Comment by deymadhumanti** [comment]
> Hi @deymadhumanti please resolve the CI failures (Spell check and prettier). And as for the breaking changes check, could you provide a justification of this? Or it is only a false alarm?
> 
> Hi @ChenTanyi would you please take a look at the Java generation failure?

Hi @ArcturusZhang , Prettier Check is successful here. I am looking at SpellCheck failure and update it .

**Comment by deymadhumanti** [comment]
> Hi @deymadhumanti please resolve the CI failures (Spell check and prettier). And as for the breaking changes check, could you provide a justification of this? Or it is only a false alarm?
> 
> Hi @ChenTanyi would you please take a look at the Java generation failure?

Hi @ArcturusZhang , Regarding breaking change failure: 

The earlier schema for CheckNameAvailability response was erronous. It was added as part of Added Swagger details for CheckNameAvailability API #5496 and OAV example validation wasn’t performed as a part of validation for the same. SDK has not been generated for the same since the swagger changes were made.

With the current PR we are fixing the response to CheckNameAvailability API and we have also validated the same using OAV spec and example validation tool.  It is in accordance with format mentioned at - https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#check-name-availability-requests

**Comment by deymadhumanti** [comment]
> > Hi @deymadhumanti please resolve the CI failures (Spell check and prettier). And as for the breaking changes check, could you provide a justification of this? Or it is only a false alarm?
> > Hi @ChenTanyi would you please take a look at the Java generation failure?
> 
> Hi @ArcturusZhang , Prettier Check and Spell Check are successful now. For Breaking changes, I have already explained. Can you please check this

**Comment by deymadhumanti** [comment]
> @deymadhumanti Yes, it could be `PrivateLinkResourceList`.
> 
> It could be anything you like which could be separated from other resources and their plurality.
> For example, the java sdk will automatically collect all `privateLinkResource` into `privateLinkResources`. And so as others.

Hi @ChenTanyi , I have changed PrivateLinkResources to PrivateLinkResourceList. But still Java SDK generation is failing with same error. Can you please take a look

**Comment by ArcturusZhang** [comment]
> > > Hi @deymadhumanti please resolve the CI failures (Spell check and prettier). And as for the breaking changes check, could you provide a justification of this? Or it is only a false alarm?
> > > Hi @ChenTanyi would you please take a look at the Java generation failure?
> > 
> > 
> > Hi @ArcturusZhang , Prettier Check and Spell Check are successful now. For Breaking changes, I have already explained. Can you please check this

I think the breaking changes are reasonable based on what you said. But since the breaking changes CI is a required check, this PR must wait for @akning-ms 's approval and eventually merge.
Hi @akning-ms could you please have a look at this?

**Comment by ChenTanyi** [comment]
@deymadhumanti After my test, it is not the error caused by items, the `privateLinkResourceList` can be changed back if you like. 

The problem is in operations, java always use operations in plural format.
[PrivateLinkResources_List](https://github.com/Azure/azure-rest-api-specs/pull/8655/files#diff-02ce4ce9148599d2d4ec2600b9c4626cR27) and [PrivateLinkResource_Get](https://github.com/Azure/azure-rest-api-specs/pull/8655/files#diff-02ce4ce9148599d2d4ec2600b9c4626cR71) in `operationId` are conflict. You can change it into same prefix, or use `directive` in readme like https://github.com/azure/azure-rest-api-specs/blob/89c0f6e94d390cd4c3bfe992bac632db6058e9fd/specification/adhybridhealthservice/resource-manager/readme.java.md#L14

**Comment by deymadhumanti** [comment]
> Hi @deymadhumanti please resolve the CI failures (Spell check and prettier). And as for the breaking changes check, could you provide a justification of this? Or it is only a false alarm?
> Hi @ChenTanyi would you please take a look at the Java generation failure?
> 
> Hi @ArcturusZhang , Prettier Check and Spell Check are successful now. For Breaking changes, I have already explained. Can you please check this
> 
> 
> I think the breaking changes are reasonable based on what you said. But since the breaking changes CI is a required check, this PR must wait for @akning-ms 's approval and eventually merge.
> Hi @akning-ms could you please have a look at this?

Hi @akning-ms, Can you please take a look at this PR.

**Pull request merged**
